proxy/src/control/destination/mod.rs

```diff
@@ -267,3 +348,56 @@ impl Metadata {
         self.tls_identity.as_ref()
     }
 }
+
+// ===== impl RebindState =====
+
+impl RebindState {
```

Comment:
First, the main goal with my (offline) suggestion today is that we find a way to get rid of `RebindState` since it seems redundant with the state we already maintain both in the load balancer and in the discovery module (`DestinationSet`), and IMO it adds quite a bit of complexity.

In this PR, `Resolver` receives a notification that the TLS configuration changed and then asynchronously sends a bunch of `Change::Insert`s to the load balancer in order to get the load balancer to gracefully disconnect the connections, i.e. invalidate its connection cache.

In the other PR, following my earlier (offline) suggestion, we have the discovery background service noticing that the TLS configuration changed; then it tells updates to the Resolver that then tells the load balancer to invalidate its connection cache. So, basically that gets rid of RebindState by adding a new layer of indirection, at the cost of making service discovery more complicated with functionality that it ideally wouldn't have, right?

Now, let me ask a dumb question: Why can't we have something watch for TLS configuration changes, and then directly tell the load balancer directly that it needs to invalidate all items that meet some criteria (namely, TLS connections using the older configuration) from its cache, so that it will create new connections? This would avoid needing to change Resolver and the Destination service completely, and avoid the need for a `RebindState` to duplicate the load balancer (and service discovery) state.

Reply:
@briansmith 
> Why can't we have something watch for TLS configuration changes, and then directly tell the load balancer directly that it needs to invalidate all items that meet some criteria (namely, TLS connections using the older configuration) from its cache, so that it will create new connections?

The load balancer doesn't know how to bind a service on its' own. Instead, it polls something implementing `tower_discover::Discover` for `Change::Insert` events with already bound `Service`s. The invalidation and rebinding can't happen entirely inside the load balancer (without significant changes to its design) for this reason. Furthermore, with the current implementation, the load balancer doesn't have the TLS metadata it would need to determine which endpoints should be rebound.

Reply:
> The invalidation and rebinding can't happen entirely inside the load balancer

Forgive me for not paying enough attention to the load balancer design until now and asking dumb questions. We're talking about `Router`, right? Note that `Router` has a cache that contains `Node`s that look like this:

```rust
/// Wraps cache values so that each tracks its last access time.
#[derive(Debug, PartialEq)]
pub struct Node<T> {
    value: T,
    last_access: Instant,
}
```

It seems like it would be straightforward to add a "is_tls" field to every `Node`, and then add a `Router::invalidate_tls()` function that calls a `Cache::invalidate_tls()` function that removes every `Node` from the cache that has `is_tls` is true, similar to how `Router::call()` calls `Cache::reserve()` to invalidate the least recently used (or whatever) entry.

Reply:
@briansmith 
> We're talking about`Router`, right?

No, we're concerned with the load balancer (`Balance`) from the `tower_balance` crate. The router is responsible for routing requests to logical _services_. Those services are each represented by a _service stack_. The outbound _service stack_s each contain a load balancer for that service. The load balancer fans out requests to multiple _endpoints_ that implement that service. Each endpoint has a _client stack_ of middleware. We want to rebind at the level of the _client_ stacks, not the _service_ stacks. The service, for the purposes of routing, is unchanged, but the details of each individual client connection (as represented by the client stack) should be rebound.

Just to make sure we're all on the same page, @olix0r, is my understanding of this correct from your perspective?

Reply:
So, if we have to choose between *this* PR and #1175, I greatly prefer #1175 because it avoids the complexity of `RebindState`, in particular the `now` vs. `next`. In #1175, we're effectively using the channel between the resolver and the discovery background task as the ``RebindState`, right?

Reply:
I think the intent that this happen beneath the load balancer is correct. I definitely do **not** think that we should maintain an additional cache here. I'll need to try to propose a patch myself to have any chance of being concrete about this.